Sources/AloeStackView/Animation/AnimationCoordinator.swift

```diff
@@ -0,0 +1,64 @@
+// Created by gwangbeom on 10/29/18.
+// Copyright 2018 Airbnb, Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+
+/**
+ * An object that animates when the Row is inserted and deleted from the AloeStackView.
+ */
+public class AnimationCoordinator {
```

Comment:
I would like to simplify this implementation by potentially changing how this works. Rather than passing an `AnimationCoordinator` into the row view, what if instead the custom animating protocol had methods such as:

    func animateInsert()
    func animateRemove()

These methods would be called from inside the animation block when the row insertion or removal happens. The view can make any changes it wants inside these methods and the changes will be animated.

Additionally, we could add some methods to the custom animating protocol to allow for the initial state to be configured before the animation, and to perform actions after the animation completes, for example:

    func setupInsertAnimation()
    func didCompleteInsertAnimation()

    func setupRemoveAnimation()
    func didCompleteRemoveAnimation()

You could then implement custom animations by doing something like:

    func setupInsertAnimation() {
      transform = CGAffineTransform(translationX: -100, y: 0)
    }

    func animateInsert() {
      transform = .identity
    }

I think this could be a lot simpler and clearer. Let me know what you think of this approach.

Reply:
Thank you for your feedback.
I also implemented it early on through the above approach.
If you implement it in the above way, there will be an advantage that the function name makes meaning clear.
But there are too many functions inside the protocol.
So to simplify this, I thought about the viewWillTransition function which is the api of UIVIewController.
Please tell me your thoughts on my opinion. 
Thank you!

Reply:
Yes you’re right, it does create too many functions in the protocol.

My plan to address that was to add a protocol extension with empty definitions for every function. That way views only need to implement the specific functions they care about and don’t need to implement every function.

For example:

    extension CustomAnimating {

      func willAnimateInsert() { }
      func animateInsert() { }
      func didAnimateInsert() { }

      ...

    }

I think this will give the best of both worlds. It also has the advantage that we can add more custom animation methods in the future, for example we can add methods for customizing show and hide animations, without having one big animation method becoming unwieldy in every view.

I also updated the names of their methods in my previous comments to better match the UIKit will/did naming convention.

Let me know what you think!

Reply:
Listening to you, I think it makes sense! 👍 

```swift
protocol CustomAnimating {

  func insertAnimationWillBegin()
  func animateInsert()
  func insertAnimationDidEnd(_ completed: Bool)

  ...
}
```
What do you think about naming such a thing?
I want to hear your thoughts.

Reply:
Yes, great suggestion! I like it much better 😄

Perhaps it might be easier to understand if we put the protocol methods in this order:

```swift
protocol CustomAnimating {

  /// ...
  func animateInsert()

  /// ...
  func insertAnimationWillBegin()

  /// ...
  func insertAnimationDidEnd(_ finished: Bool)

  ...

}
```

That way people will see the most important method in the protocol first (the one they need to implement to perform the custom animation).

Also, perhaps the argument to `insertAnimationDidEnd` should be named `finished` rather than `completed` so that it matches the name UIKit uses in `animate(withDuration:animations:completion:)` (the completion callback takes an argument named `finished`).

Let me know your thoughts!

Reply:
I agree with your thoughts. 👍